File: 진우/16-17weeks/25501. 재귀의 귀재/재귀의 귀재.cc
#include <iostream>   
#include <string>     

using namespace std;

// 재귀 함수가 호출된 횟수를 세기 위한 전역 변수
int cnt = 0;

// 문자열 s의 l(왼쪽)부터 r(오른쪽)까지가 팰린드롬인지 재귀적으로 검사하는 함수
int recursion(const string& s, int l, int r) {
    cnt++;  // recursion 함수가 호출될 때마다 카운트 증가

    // 종료 조건 1: 중앙까지 도달한 경우 (팰린드롬 성립)
    if (l >= r) {
        return 1;
    }
    // 종료 조건 2: 양쪽 문자가 다르면 팰린드롬이 아님
    else if (s[l] != s[r]) {
        return 0;
    }
    // 양쪽 문자가 같으면 한 칸씩 안쪽으로 이동하여 재귀 호출
    else {
        return recursion(s, l + 1, r - 1);
    }	
}

// 문자열 전체에 대해 팰린드롬 여부를 검사하는 함수
int isPalindrome(const string& s) {
    cnt = 0;  // 새로운 문자열 검사 전에 재귀 호출 횟수 초기화
    // 문자열의 시작(0)부터 끝(length - 1)까지 검사
    return recursion(s, 0, s.length() - 1);
}

int main() {
    int N;
    cin >> N;  // 검사할 문자열 개수 입력

    string word;
    // N개의 문자열을 하나씩 입력받아 팰린드롬 여부 검사
    for (int i = 0; i < N; i++) {
        cin >> word;  // 문자열 입력
        // 팰린드롬 결과(1 또는 0)와 재귀 호출 횟수 출력
        cout << isPalindrome(word) << " " << cnt << "\n";
    }

    return 0;  
}

File: 진우/16-17weeks/25501. 재귀의 귀재/재귀의 귀재.py
# 재귀 함수가 호출된 횟수를 저장하는 전역 변수
cnt = 0

# 문자열 s의 양 끝(l, r)을 비교하며 팰린드롬 여부를 재귀적으로 판단하는 함수
def recursion(s, l, r):
    global cnt          # 전역 변수 cnt를 사용하겠다고 선언
    cnt += 1            # recursion 함수가 호출될 때마다 카운트 증가

    # 종료 조건 1: 왼쪽 인덱스가 오른쪽 인덱스를 넘거나 같아지면
    # 문자열의 중앙에 도달했으므로 팰린드롬 성립
    if l >= r:
        return 1

    # 종료 조건 2: 양쪽 문자가 다르면 팰린드롬이 아님
    elif s[l] != s[r]:
        return 0

    # 양쪽 문자가 같으면 한 칸씩 안쪽으로 이동하여 다시 검사
    else:
        return recursion(s, l + 1, r - 1)

# 문자열 전체에 대해 팰린드롬 여부를 검사하는 함수
def isPalindrome(s):
    global cnt          # 전역 cnt를 초기화하기 위해 선언
    cnt = 0             # 새로운 문자열 검사 전에 호출 횟수 초기화
    # 문자열의 처음(0)부터 끝(len(s) - 1)까지 검사 시작
    return recursion(s, 0, len(s) - 1)

# 테스트 케이스 개수 입력
T = int(input())

# T개의 문자열에 대해 팰린드롬 검사 수행
for _ in range(T):
    word = input()      # 검사할 문자열 입력

    # 팰린드롬 여부(1 또는 0)와 재귀 호출 횟수 출력
    print(isPalindrome(word), cnt)
